### PR TITLE
Added anvil gui to make custom withdraw/deposit actions via gui possible

### PR DESCRIPTION
**Pull request by TheBlackEntity**
I did test it for 1.17.1, but haven't tested it for any other versions, yet, but it should work...

I'm using the VirtualAnvil Code in a project of mine and worked so far.

Here's the source:
https://github.com/TheBlackEntity/VirtualAnvil

Would be nice if you added this feature by using my source code, or just using your own

### CODE REVIEW

File: src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+        if (!(e.getClickedInventory() instanceof AnvilInventory))
+            return;
+
+        AnvilInventory anvilInventory = (AnvilInventory) e.getClickedInventory();
+
+        if (e.getSlot() == 2) {
+            if (e.getWhoClicked().getExpToLevel() >= anvilInventory.getRepairCost())
+                ((Player) e.getWhoClicked()).giveExpLevels(0);
+            Data data = this.opened.get(e.getWhoClicked().getUniqueId());
```
Comment (DoctaEnkoda) [inline src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java:61]: Ligne 48 :
Do not do the same search twice, retrieve the value at the beginning and check if it is null.

File: src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            e.setCancelled(true);
```
Comment (DoctaEnkoda) [inline src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java:62]: 2 times cancelled ??

File: src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.*;
+
+public class IslandCustomAmountBankGUI implements Listener {
+    private final Map<UUID, Data> opened = new HashMap();
+
+    public IslandCustomAmountBankGUI() {
+        if (VirtualAnvil.getVirtualAnvil() == null)
+            VirtualAnvil.load();
+
+        Bukkit.getPluginManager().registerEvents(this, IridiumSkyblock.getInstance());
```
Comment (DoctaEnkoda) [inline src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java:28]: not here, in IridiumSkyblock.java

File: src/main/java/com/iridium/iridiumskyblock/gui/IslandBankGUI.java [hunk 4]
```diff
@@ -31,16 +32,15 @@ public IslandBankGUI(@NotNull Island island, Inventory previousInventory) {
...
         for (BankItem bankItem : IridiumSkyblock.getInstance().getBankItemList()) {
-            IslandBank islandBank = IridiumSkyblock.getInstance().getIslandManager().getIslandBank(getIsland(), bankItem);
+            IslandBank islandBank = IridiumSkyblock.getInstance().getIslandManager().getIslandBank(this.getIsland(), bankItem);
             inventory.setItem(bankItem.getItem().slot, ItemStackUtils.makeItem(bankItem.getItem(), Collections.singletonList(new Placeholder("amount", IridiumSkyblock.getInstance().getNumberFormatter().format(islandBank.getNumber())))));
         }
 
-        if (IridiumSkyblock.getInstance().getConfiguration().backButtons && getPreviousInventory() != null) {
+        if (IridiumSkyblock.getInstance().getConfiguration().backButtons && this.getPreviousInventory() != null)
             inventory.setItem(inventory.getSize() + IridiumSkyblock.getInstance().getInventories().backButton.slot, ItemStackUtils.makeItem(IridiumSkyblock.getInstance().getInventories().backButton));
-        }
```
Comment (DoctaEnkoda) [inline src/main/java/com/iridium/iridiumskyblock/gui/IslandBankGUI.java:43]: Restore

File: src/main/java/com/iridium/iridiumskyblock/commands/DepositCommand.java [hunk 5]
```diff
@@ -80,9 +82,16 @@ public boolean execute(CommandSender sender, String[] args) {
...
+        if (args.length == 2) {
+            List<String> possible = IridiumSkyblock.getInstance().getBankItemList().stream()
+                    .map(BankItem::getName)
+                    .collect(Collectors.toList());
+
+            List<String> actualTab = possible.stream().filter(s -> s.toLowerCase(Locale.ROOT).startsWith(args[1].toLowerCase(Locale.ROOT))).collect(Collectors.toList());
+            return actualTab.isEmpty() ? possible : actualTab;
+        }
+
+        return Collections.singletonList("");
```
Comment (DoctaEnkoda) [inline src/main/java/com/iridium/iridiumskyblock/commands/DepositCommand.java:94]:
```suggestion
        return Collections.emptyList();
```

File: src/main/java/com/iridium/iridiumskyblock/bank/CrystalsBankItem.java [hunk 6]
```diff
@@ -51,20 +51,18 @@ public double withdraw(Player player, Number amount) {
...
-                        .replace("%type%", getDisplayName())
+                                .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix))
+                        .replace("%type%", this.getDisplayName())
                 );
-            }
             return crystals;
-        } else {
+        } else
             player.sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().noIsland.replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix)));
-        }
```
Comment (DoctaEnkoda) [inline src/main/java/com/iridium/iridiumskyblock/bank/CrystalsBankItem.java:67]: Restore

File: src/main/java/com/iridium/iridiumskyblock/bank/CrystalsBankItem.java [hunk 7]
```diff
@@ -51,20 +51,18 @@ public double withdraw(Player player, Number amount) {
...
                 );
                 player.sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().bankWithdrew
-                        .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix))
+                                .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix))
                         .replace("%amount%", String.valueOf(crystals))
-                        .replace("%type%", getDisplayName())
+                        .replace("%type%", this.getDisplayName())
                 );
-            } else {
+            } else
```
Comment (DoctaEnkoda) [inline src/main/java/com/iridium/iridiumskyblock/bank/CrystalsBankItem.java:58]: Restore

**Review by DoctaEnkoda — CHANGES_REQUESTED** [review]
Clearly, I haven't tested the code, but looking like this, there are quite a few changes to be made

File: src/main/java/com/iridium/iridiumskyblock/IridiumSkyblock.java [hunk 8]
```diff
@@ -233,6 +235,8 @@ public void run() {
             }
         }
 
+        this.islandCustomAmountBankGUI = new IslandCustomAmountBankGUI();
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/IridiumSkyblock.java:238]: I dont like this, why dont we just create the new class when needed like we do everywhere else. I see no reason to store it here

File: src/main/java/com/iridium/iridiumskyblock/gui/IslandBankGUI.java [hunk 9]
```diff
@@ -54,6 +54,22 @@ public void onInventoryClick(InventoryClickEvent event) {
         Optional<BankItem> bankItem = IridiumSkyblock.getInstance().getBankItemList().stream().filter(item -> item.getItem().slot == event.getSlot()).findFirst();
         if (!bankItem.isPresent()) return;
 
+        if (event.isShiftClick()) {
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/gui/IslandBankGUI.java:57]: This is kinda weird, and you didnt delete the old code that was there origionally?

File: src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            } catch (NullPointerException | NumberFormatException ignored) {
+                e.getWhoClicked().sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().notANumber.replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix)));
+            }
+
+            e.getWhoClicked().openInventory(e.getInventory());
+            e.getWhoClicked().closeInventory();
+        }
+    }
+
+    private static final class Data {
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java:80]: Should be using Lombok's @Getter and @AllArgsConstructor here

**Comment by PeachesMLG** [comment]
Also need to adress failing unit tests

**Comment by dlsf** [comment]
I'll be blunt here: I don't like this PR. It introduces an additional dependency but still has a lot of mediocre code. The idea is good, though.

**Comment by dlsf** [comment]
I also don't really see the point of adding `this.` everywhere, we usually don't use it unless it's really necessary or greatly improves readability

**Comment by TheBlackEntity** [comment]
> I also don't really see the point of adding `this.` everywhere, we usually don't use it unless it's really necessary or greatly improves readability

It's added by a plugin of my ide

**Comment by TheBlackEntity** [comment]
> I'll be blunt here: I don't like this PR. It introduces an additional dependency but still has a lot of mediocre code. The idea is good, though.

I know the code isn't exactly what you call perfect, but it works

File: src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java (hunk 10 above)
Comment (TheBlackEntity) [reply src/main/java/com/iridium/iridiumskyblock/gui/IslandCustomAmountBankGUI.java:80]: Haven't worked much with Lombok, yet.

Will read into it though

**Pull request closed without merging**